packages/xo-server/src/xo-mixins/proxies.js

```diff
@@ -60,48 +69,168 @@ export default class Proxy {
-    if (address !== undefined && proxy.address !== address) {
-      await this._throwIfRegistered(address)
-    }
+  async updateProxy(id, { address, authenticationToken, name, vmUuid }) {
+    await this._throwIfRegistered(address, vmUuid)
 
-    patch(proxy, { address, authenticationToken, name })
-    return this._db.update(proxy).then(extractProperties)
+    const proxy = await this._getProxy(id)
+    patch(proxy, { address, authenticationToken, name, vmUuid })
+    return this._db
+      .update(proxy)
+      .then(extractProperties)
+      .then(omitToken)
   }
 
-  async callProxyMethod(id, method, params, expectStream = false) {
-    const proxy = await this.getProxy(id)
-
-    const response = await timeout.call(
-      this._app.httpRequest(proxy.address, {
-        body: format.request(0, method, params),
-        headers: {
-          'Content-Type': 'application/json',
-          Cookie: cookie.serialize(
-            'authenticationToken',
-            proxy.authenticationToken
-          ),
-        },
-        method: 'POST',
-        pathname: '/api/v1',
+  async upgradeProxyAppliance(id) {
+    const { vmUuid } = await this._getProxy(id)
+    const xapi = this._app.getXapi(vmUuid)
+    await xapi.getObject(vmUuid).update_xenstore_data({
+      'vm-data/xoa-updater-channel': JSON.stringify(this._xoProxyConf.channel),
+    })
+
+    return xapi.rebootVm(vmUuid, { hard: true })
+  }
+
+  @defer
+  async deployProxy($defer, srId) {
+    const app = this._app
+    const xoProxyConf = this._xoProxyConf
+
+    const namespace = xoProxyConf.namespace
+    const {
+      [namespace]: { xva },
+    } = await app.getResourceCatalog()
+    const xapi = app.getXapi(srId)
+    let vm = await xapi.importVm(
+      await app.requestResource({
+        id: xva.id,
+        namespace,
+        version: xva.version,
+      }),
+      { srId }
+    )
+    $defer.onFailure.call(xapi, '_deleteVm', vm)
+
+    const [
+      password,
+      token,
+      { registrationToken, registrationEmail: email },
+    ] = await Promise.all([
+      generateToken(10),
+      generateToken(),
+      app.getApplianceRegistration(),
+    ])
+    const date = new Date().toISOString()
+    await Promise.all([
+      vm.add_tags('XOA Proxy'),
+      vm.set_name_label(`XOA Proxy ${date}`),
+      vm.update_xenstore_data({
+        'vm-data/system-account-xoa-password': password,
+        'vm-data/xo-proxy-authenticationToken': JSON.stringify(token),
+        'vm-data/xoa-updater-credentials': JSON.stringify({
+          email,
+          registrationToken,
+        }),
+        'vm-data/xoa-updater-channel': JSON.stringify(xoProxyConf.channel),
       }),
-      parseDuration(this._xoProxyConf.callTimeout)
+    ])
+
+    await xapi.startVm(vm.$id)
+
+    await vm.update_xenstore_data({
+      'vm-data/system-account-xoa-password': null,
+      'vm-data/xo-proxy-authenticationToken': null,
+      'vm-data/xoa-updater-credentials': null,
+    })
```

Comment:
Didn't you say this didn't work?

Reply:
It works, it was just a non-updated entry.